api/app/signals/apps/signals/models/category_assignment.py

```diff
@@ -17,6 +18,16 @@ class CategoryAssignment(CreatedUpdatedModel):
 
     extra_properties = JSONField(null=True)  # TODO: candidate for removal
 
+    # Note on the deadline fields: nulls are allowed for historic data because
+    # historic service level objectives are not generally available.
+    deadline = models.DateTimeField(null=True)
+    deadline_factor_3 = models.DateTimeField(null=True)
+
     def __str__(self):
         """String representation."""
         return '{sub} - {signal}'.format(sub=self.category, signal=self._signal)
+
+    def save(self, *args, **kwargs):
+        self.deadline, self.deadline_factor_3 = DeadlineCalculationService.from_signal_and_category(
+            self._signal, self.category)
```

Comment:
Do we want this calculation to be run every time we save the `CategoryAssingment`? Maybe add a check if the `deadline` and `deadline_factor_3` are already set to prevent this calculation to overwrite the values every time a save is done. Not a blocker because the `CategoryAssingment` should not be saved a lot. On the other hand this will make sure the deadlines are always correct even if I manually alter them and save 🤔 .

Reply:
This only happens on a write, I think making sure the numbers are correct is worth it. I think optimizing the read side of our API is more worth it (happens much more often).

So I will leave this.